iup/__init__.py

```diff
@@ -0,0 +1,361 @@
+import polars as pl
+import datetime as dt
+from sklearn.linear_model import LinearRegression
+import numpy as np
+import abc
+from typing_extensions import Self
+
+
+class UptakeData(pl.DataFrame, metaclass=abc.ABCMeta):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.validate()
+
+    @abc.abstractmethod
+    def validate(self) -> None:
+        pass
+
+    def assert_columns_found(self, columns):
+        for col in columns:
+            assert col in self.columns, f"Column {col} is expected but not found."
+
+    def assert_columns_type(self, columns, dtype):
+        for col in columns:
+            assert (
+                self[col].dtype == dtype
+            ), f"Column {col} should be {dtype} but is {self[col].dtype}"
+
+    @staticmethod
+    @abc.abstractmethod
+    def split_train_test(uptake_data_list, start_date, side):
+        pass
+
+
+class IncidentUptakeData(UptakeData):
+    def validate(self):
+        self.assert_columns_found(
+            ["region", "season", "date", "elapsed", "interval", "estimate"]
+        )
+        self.assert_columns_type(["date"], pl.Date)
+        self.assert_columns_type(["region", "season"], pl.Utf8)
+        self.assert_columns_type(["elapsed", "interval", "estimate"], pl.Float64)
+
+    @staticmethod
+    def split_train_test(uptake_data_list, start_date, side):
+        if side == "train":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") < start_date)
+            )
+        if side == "test":
+            out = (
+                pl.concat(uptake_data_list)
+                .sort("date")
+                .filter(pl.col("date") >= start_date)
+            )
+
+        out = IncidentUptakeData(out)
+
+        return out
+
+    def get_previous_uptake(self) -> Self:
+        self = self.with_columns(previous=pl.col("estimate").shift(1).over("region"))
+
+        return self
```

Comment:
Pylance flags this as an error, saying:  Type "DataFrame" is not assignable to return type "Self@IncidentUptakeData"

I don't understand this error. `get_previous_uptake` is quite analogous to `trim_outlier_intervals` in that it performs some data frame manipulation on an IncidentUptakeData object and returns it as the same object type. Why is this a problem in `get_previous_uptake` but not in `trim_outlier_intervals`?

Reply:
As far as I can tell, a polars method (e.g. `with_columns`) will always return a polars dataframe, even if it was given a custom subclass of a polars dataframe. To prevent this, I had to define my own `with_columns` method that does what polars `with_columns` does but then returns whatever subclass it was given.